src/osd/ReplicatedBackend.cc

```diff
@@ -1665,24 +1693,44 @@ void ReplicatedBackend::submit_push_data(
   map<string, bufferlist> &omap_entries,
   ObjectStore::Transaction *t)
 {
-  hobject_t target_oid;
-  if (first && complete) {
-    target_oid = recovery_info.soid;
-  } else {
-    target_oid = get_parent()->get_temp_recovery_object(recovery_info.version,
+  hobject_t target_oid = get_parent()->get_temp_recovery_object(recovery_info.version,
```

Comment:
This is a useful optimization.  For small objects (up to 8MB by default I think), we can write atomically in place instead of writing to a temp object.  I don't want to get rid of that.  When we are recovering partial extents, you can just atomically overwrite the dirty portions, no need to clone_range the clean parts into a temp object.